src/gsModeling/gsSurfaceReparameterization.h

```diff
@@ -0,0 +1,205 @@
+	for (index_t i = 0; i < n; ++i) {
+	  tmp[i] += h;
+	  const T e1 = this->evalObj(ctmp);
+	  tmp[i] = u[i] - h;
+	  const T e2 = this->evalObj(ctmp);
+	  tmp[i] = u[i];
+	  result[i] = (e1 - e2) / (2 * h);
+	}
+  }
+
+  template<short_t d, typename T>
+  void gsObjFuncSurface<d, T>::defaultOptions() {
+	m_options.addReal("qi_lambda1", "Sets the lambda 1 value", 1.0);
+	m_options.addReal("qi_lambda2", "Sets the lambda 2 value", 1.0);
+  }
+
+  template<short_t d, typename T>
+  void gsObjFuncSurface<d, T>::addOptions(const gsOptionList &options) {
+	m_options.update(options, gsOptionList::addIfUnknown);
+  }
+
+  template<short_t d, typename T>
+  void gsObjFuncSurface<d, T>::applyOptions(const gsOptionList &options) {
+	m_options.update(options, gsOptionList::addIfUnknown);
+	m_lambda1 = m_options.getReal("qi_lambda1");
+	m_lambda2 = m_options.getReal("qi_lambda2");
+	m_evaluator.options().update(m_options, gsOptionList::addIfUnknown);
+  }
+
+  template<class T = real_t>
+  gsMultiPatch<T> convertIntoBSpline(const gsMultiPatch<T>& mp, const gsMatrix<T>& coefsMobiusIn) {
+	gsMultiPatch<T> result;
+
+	// Precompute Mobius domain matrix (use structured bindings in C++17)
+	gsAsConstVector<T> coefsMobius(coefsMobiusIn.data(), 4);
+	gsMatrix<T, 2, 2> alpha;
+	alpha << coefsMobius(0), coefsMobius(2), coefsMobius(1), coefsMobius(3);
+	gsMobiusDomain<2, T> mobiusDomain(alpha);
+
+	for (const auto& patch : mp.patches()) {
+	  // Generate UV grid points for parameterization
+	  gsMatrix<T> uv = gsPointGrid(mp.parameterRange(0), patch->basis().size() * 2);
+
+	  // Evaluate the Mobius domain mapping
+	  gsMatrix<T> xieta;
+	  mobiusDomain.eval_into(uv, xieta);
+
+	  // Evaluate geometry
+	  gsMatrix<T> eval_geo(3, uv.cols());
+	  patch->eval_into(xieta, eval_geo);
+
+	  // Convert the patch's basis into a tensor B-spline basis
+	  auto& bbasis = static_cast<gsTensorBSplineBasis<2, T>&>(patch->basis());
+
+	  // Fit surface using gsFitting and adjust parameters
+	  gsFitting<T> fittingSurface(uv, eval_geo, bbasis);
+	  fittingSurface.compute();
+	  fittingSurface.parameterCorrection();
+
+	  // Add the fitted patch to the result multipatch
+	  result.addPatch(*fittingSurface.result());
+	}
+
+	return result;
+  }
+
+  template <typename T>
+  class SurfaceReparameterization {
+   public:
+	// Constructor takes a multi-patch input and alpha matrix for the Mobius domain
+	explicit SurfaceReparameterization(const gsMultiPatch<T>& patches)
+		: m_mp(patches) {
+	  gsMatrix<T, 2, 2> alpha;
+	  alpha.setConstant(0.5);
+	  m_mobiusDomain = gsMobiusDomain<2, T>(alpha);
+	}
+
+	// Run the optimization process and generate the reparameterized B-Spline surface
+	gsMultiPatch<T> solve() {
+	  gsObjFuncSurface<2, T> objFuncSurface(m_mp, m_mobiusDomain);
+
+	  gsVector<T> initialGuessVector(4);
+	  initialGuessVector.setConstant(0.5); // Adjust this as necessary for better performance
+
+	  // Set up the optimizer
+	  gsHLBFGS<T> optimizer(&objFuncSurface);
```

Comment:
Can we make the optimizer a member of the class (a pointer to gsOptimizer) or a template? In that way, we can provide the freedom to the user to select the optimizer and we don't have to disable the class without HLBFGS. 
The optimizer can be given as an argument to the constructor, and by default giving gradient descent.

Reply:
We did it here:
https://github.com/gismo/gismo/blob/b275730cf650745ebca6c845ebf3c1ff6b0b0f61/src/gsAssembler/gsAdaptiveParametrization.h#L139